packages-exp/messaging-exp/src/helpers/logToFirelog.ts

```diff
@@ -0,0 +1,215 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  DEFAULT_BACKOFF_TIME_MS,
+  EVENT_MESSAGE_DELIVERED,
+  FCM_LOG_SOURCE,
+  LOG_INTERVAL_IN_MS,
+  MAX_NUMBER_OF_EVENTS_PER_LOG_REQUEST,
+  MAX_RETRIES,
+  MessageType,
+  SDK_PLATFORM_WEB
+} from '../util/constants';
+import {
+  FcmEvent,
+  LogEvent,
+  LogRequest,
+  LogResponse
+} from '../interfaces/logging-types';
+
+import { MessagePayloadInternal } from '../interfaces/internal-message-payload';
+import { MessagingService } from '../messaging-service';
+
+const FIRELOG_ENDPOINT = _mergeStrings(
+  'hts/frbslgigp.ogepscmv/ieo/eaylg',
+  'tp:/ieaeogn-agolai.o/1frlglgc/o'
+);
+
+const FCM_TRANSPORT_KEY = _mergeStrings(
+  'AzSCbw63g1R0nCw85jG8',
+  'Iaya3yLKwmgvh7cF0q4'
+);
+
+export function startLoggingService(messaging: MessagingService): void {
+  if (!messaging.isLogServiceStarted) {
+    _processQueue(messaging, LOG_INTERVAL_IN_MS);
+    messaging.isLogServiceStarted = true;
+  }
+}
+
+export function _processQueue(
+  messaging: MessagingService,
+  offsetInMs: number
+): void {
+  setTimeout(async () => {
+    if (!messaging.deliveryMetricsExportedToBigQueryEnabled) {
+      // flush events and terminate logging service
+      messaging.logEvents = [];
+      messaging.isLogServiceStarted = false;
+
+      return;
+    }
+
+    if (!messaging.logEvents.length) {
+      return _processQueue(messaging, LOG_INTERVAL_IN_MS);
+    }
+
+    await _dispatchLogEvents(messaging);
+  }, offsetInMs);
+}
+
+export async function _dispatchLogEvents(
+  messaging: MessagingService
+): Promise<void> {
+  for (
+    let i = 0, n = messaging.logEvents.length;
+    i < n;
+    i += MAX_NUMBER_OF_EVENTS_PER_LOG_REQUEST
+  ) {
+    const logRequest = _createLogRequest(
+      messaging.logEvents.slice(i, i + MAX_NUMBER_OF_EVENTS_PER_LOG_REQUEST)
+    );
+
+    let retryCount = 0,
+      response = {} as Response;
+
+    do {
+      try {
+        response = await fetch(
+          FIRELOG_ENDPOINT.concat('?key=', FCM_TRANSPORT_KEY),
+          {
+            method: 'POST',
+            body: JSON.stringify(logRequest)
+          }
+        );
+
+        break;
+      } catch (error) {
+        const isLastAttempt = retryCount === MAX_RETRIES;
+        if (isLastAttempt) {
+          break;
+        }
+      }
+
+      let delayInMs: number;
+      try {
+        delayInMs = Number(
+          ((await response.json()) as LogResponse).nextRequestWaitMillis
+        );
+      } catch (e) {
+        delayInMs = DEFAULT_BACKOFF_TIME_MS;
+      }
+
+      await new Promise(resolve => setTimeout(resolve, delayInMs));
```

Comment:
In remote config, we add random fuzz to the retry delays to avoid waves of retries at exact the same time. Maybe worth applying the same here. 

https://github.com/firebase/firebase-js-sdk/blob/7028c1159b9153f14c22a81fdd09df68ffcfcfb6/packages/util/src/exponential_backoff.ts#L63

Reply:
yeah, it sounds unlikely. I think fuzzing might reduce the chance even further, but it's probably fine as is.